### DISCUSSION TITLE
Nuxt 3 and Axios

### DISCUSSION THREAD

**Discussion by KahrLu** [nuxt/nuxt#16605] [Ideas]
Hi @pi0 ,I would like to I solemnly suggest nuxt3 supporting Axios as soon as possible.

At present, nuxt3 has reached the RC stage, but it does not support Axios, which makes us worried about going online the website developed with a stable version of nuxt3 in June

According to the latest documents, official seems to encourage the use of their own ohmyfect and does not intend to make nuxt3 support Axios.
![image](https://user-images.githubusercontent.com/76502123/164582940-a159c110-baa5-491c-a3dc-6bd65d9dcbeb.png)
But as we all know that Axios is the most commonly used HTTP request Library in front-end engineering, which has amazing 92.6k stars!
It is not only stable and verified by a large number of production environments, but also one of the development habits of most Vue developers.

In fact, it is not very difficult for this plug-in to support vue3 and nuxt3. Relevant third-party repos have appeared in the community.
such as https://github.com/Teranode/nuxt-module-alternatives

Finally, we hope that official support will be given to the use of Axios in nuxt3.

note: If your company or individual has the same demand, please give the thumbs-up

**Comment by pi0** [comment]
Hi @KahrLu. Indeed it is not that hard to make [axios](https://github.com/axios/axios) _just working with nuxt3_ and I suppose it should already work if you import it! (see *) But please consider that Axios is built on top of very legacy API with `XMLHttpRequest` and Node.js API that cannot run in the workers easily. We are doing everything to move projects to leverage native [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) that is the modern and universal way to fetch in web applications.

What are the main limitations stopping you from migrating from axios to `fetch` or `$fetch`?

---

(*) I made Axios module as a wrapper to axios library because even from the beginning axios had serious issues with SSR like context sharing with headers. Features like default base URL set to `localhost:3000` are not even relevant anymore as whole Nuxt 3 stack is based on native fetch and local fetch that doesn't need a base URL anymore! Global options setting was also a design pattern with global service injections of `$axios` while when migrating to Vue 3 composition API, you can easily create them using a factory function and sub-feature of proxy is to be directly added to nitro server. I will probably add basic support to ease the migration with runtime warning since axios is far most downloaded module but it simply doesn't fits into modern setup of Nuxt 3.

**Reply by KahrLu** [reply]
Thank you for your professional reply:)
1、I knew serious issues with SSR like context sharing with headers, but we don't always need to face it directly. Besides, can't we avoid it?
2、relevant anymore as whole Nuxt 3 stack is based on native fetch and local fetch that doesn't need a base URL anymore:
     In fact, I believe at least half of the projects will not use nuxt as the full stack, and many use other back-end languages, such as Java, go and so on.That's also my current situation,  So maybe it's a one-sided statement that we no longer need baseurl.
3、It is certainly wrong to deny ohmyfetch just because Axios gained many stars, but now the ratio is 90k:900. Although I agree with some advanced ideas of ohmyfetch as I am a developer, please understand us that we can't convince an enterprise project to adopt it right now

Based on the above reasons, I think it is necessary to upgrade nuxt/Axios & nuxt/proxy modules as soon as possible to adapt to vue3. At least we should allow developers to choose freely in ohmyfetch and Axios. 

I believe nuxt3 is a project that embraces all rivers^_^ That's why we chose it，thanks!

**Reply by pi0** [reply]
> It is certainly wrong to deny ohmyfetch just because Axios gained many stars, but now the ratio is 90k:900. Although I agree with some advanced ideas of ohmyfetch as a developer, please understand us that we can't convince an enterprise project to adopt it right now

`$fetch` ([unjs/ohmyfetch](https://github.com/unjs/ohmyfetch)) is basically a syntax sugar over native [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) just like a helper utility. You are trusting an industry-standard API supported natively in all modern browsers and server environments. You can always choose that or another library on top of fetch such as [ky](https://www.npmjs.com/package/ky). It is much more stable than technology axios is using behind the scene. 


> I believe nuxt3 is a project that embraces all rivers

It is what we always truly believe. But we will also never embrace rivers in a wrong direction and guide Nuxt users with latest best practices. Just like Jquery and Lodash, libraries and technologies have an expiration date and I believe Axios is already expired for modern web era. We are pushing Nuxt and Nuxt users to the best standards such as ESM and Fetch API while everything else is free to choose. ohmyfetch is never enforced you can choose `fetch`. 

> In fact, I believe at least half of the projects will not use nuxt as the full stack, and many use other back-end languages, such as Java, go and so on.That's also my current situation, So maybe it's a one-sided statement that we no longer need baseurl.

For sure about using external backend. I mentioned it because it is one of axios (module) main features to have global base URL. We can create a [composable](https://v3.nuxtjs.org/guide/directory-structure/composables) that creates fetch with any external baseURL.


**Appendix**

With all of that said, I've no strong opinions if you really want to keep using Axios library at your own risk in Nuxt projects. We will eventually also add support for Nuxt module but it would be **only and only** to give a migration phase with strong advice to migrate into fetch API. Whatever library you choose.

Here is an example you can still use axios in Nuxt 3 with 5 lines of code: https://stackblitz.com/edit/github-qvegun-eyrptb?file=app.vue

**Reply by Aareksio** [reply]
`$fetch` is to `axios` as strange, as `nuxt 3` is to `nuxt 2`, despite the name - if you managed to convince enterprise to jump on completely different architecture of nuxt 3, you can do the same with $fetch. That said, there is a price to pay when migrating - starting with the proxy module, which AFAIK can not even be replicated in nuxt 3 in current state (local fetch mocks streams used for proxying when done during SSR).

**Reply by pi0** [reply]
@Aareksio Local fetch is never involved in a proxy setup. It is only for URLs starting with `/`, not external URLs and Nuxt 3 is still evolving. I would be more than happy to hear your feedback about missing features we can add to complete this migration.

**Reply by KahrLu** [reply]
Okay, so I still want to make sure whether nuxt/axios module will adapt to vue3 before or at the same time nuxt3 releases the stable version, which is a key time node.
and nuxt/proxy, from you two gentlemen 'comments, no matter $fetch or Axios, we don't have a proxy to support accessing external APIs at present, right?

**Reply by pi0** [reply]
Built-in proxy support => Soon! (https://github.com/unjs/nitro/issues/113)

Axios => Believe it or not, you can already use axios with Vue 3 and Nuxt 3 :) (see stackblitz link above). What features do you think might be missing other than proxy that module support will add?

**Reply by KahrLu** [reply]
@pi0 awesome! will flow nitro, too.
That's all I want to know, and I believe we can use Axios in nuxt3 immediately, because that's what I have done.
I haven't found any other functions that need to be added yet. Thank you

**Comment by Aareksio** [comment]
@pi0 Following on your comment, what is the recommended approach to handling API available on port `4000` both during dev and in deployment (node + SSR on https://example.com):

**dev**
Server: `$fetch('/api')` -> `http://localhost:4000/api`
Client: `$fetch('/api')` -> `http://localhost:4000/api`

**deployment**
Server: `$fetch('/api')` -> `http://localhost:4000/api`
Client: `$fetch('/api')` -> `https://example.com/api` -> (lb, reverse proxy) -> `http://localhost:4000/api`

Is the solution coming in [nitro#113](https://github.com/unjs/nitro/issues/113)?

@KahrLu As you said in edit, I too believe we do not have sufficient proxy support. There is no documentation on that. But the fact is, no one have actually asked for the solution yet :)

**Reply by KahrLu** [reply]
At present, I directly specify baseurl as the host of the external API and authorize cross domain on the server.  wait for [nitro#113](https://github.com/unjs/nitro/issues/113?)

**Comment by ildan95** [comment]
Hello. I try to use fetch instead of axios, but I got unexpected behavior. When I do useFetch second time to the same url, no new request is created.  In my case, I send first time, got 400 error with some my data. Then send one more time to the same url and got "true" in error but without my data (request not sent). I didn't find how to disable this behavior. I need to send new same request. Like in axios. Can anyone help me?

**Reply by ildan95** [reply]
I found solution. set option "initialCache" to false in useFetch options. But it isn't explicit behavior. I think it have to be false as default

**Reply by pi0** [reply]
As an additional tip, you can always use `$fetch` or `fetch` for more customized behavior.

**Comment by viii47** [comment]
Just wanted to note that axios works right out of the box for the ~/server structure already in Nuxt3. I currently use it to proxify/chain/resolve formData type to a handful of backend systems.

Example
```ts
import { API_HTTP_OSM } from "~/server/constants/index";
import axios from "axios";

export default defineEventHandler(async (event) => {
    const body = await useBody(event);
    console.log(body);

    const type: "QUERY" | "STREET" = body.options.type;
    let query = `q=${encodeURIComponent(body.options.query.toString())}`;
    const data = axios({
        url: `${API_HTTP_OSM}/search?${query}&format=jsonv2&addressdetails=1&limit=5`,
        method: "GET",
    })
        .then((response) => {
            return response.data;
        })
        .catch((error) => {
            return error;
        });

    return data;
});

```

**Reply by pi0** [reply]
You can also do this without axios using native fetch:

```js
import { API_HTTP_OSM } from '~/server/constants/index'

export default defineEventHandler(async (event) => {
  const body = await useBody(event)
  // const type: "QUERY" | "STREET" = body.options.type;
  const data = await $fetch(`${API_HTTP_OSM}/search`, {
    params: {
      q: encodeURIComponent(body.options.query.toString()),
      format: 'jsonv2',
      addressdetails: '1',
      limit: '5'
    }
  }).catch(error => error)
  return data
})
```

**Reply by viii47** [reply]
That's true. Just wanted to show that axios is already functioning :)

**Comment by Kasheftin** [comment]
I made several nuxt2 projects in the past and now experience problems trying to start with nuxt3.

Fetch in nuxt2 is not $fetch in nuxt3. It was helper for page-level components that allowed to make requests (with axios) universally both on client and server. Usually you make a request and dispatch the response to vuex. Or even the vuex action can call axios by itself, then fetch is used just for triggering the corresponding vuex action. Now $fetch is tightly coupled with the url. 

**Is there any old-fetch-nuxt2 helper for page level components in nuxt3?**

Using $fetch is not convenient because it's not type safe.

**Is there any openapi -> nuxt3 $fetch code generator?**

That's how nuxt2 works with openapi: swagger generates api.ts using axios template. Front-end gets a file like:

````javascript
import { AxiosInstance, AxiosRequestConfig } from 'axios'

export class FilesClient {
  private instance: AxiosInstance;
  private baseUrl: string;

  constructor(baseUrl: string, instance: AxiosInstance) {
    this.instance = instance
    this.baseUrl = baseUrl
  }

  uri(id: string , cancelToken?: CancelToken | undefined): Promise<string> {
    let url_ = this.baseUrl + "/api/internal/files/uri/{id}";
    if (id === undefined || id === null)
      throw new Error("The parameter 'id' must be defined.");
    url_ = url_.replace("{id}", encodeURIComponent("" + id));
    url_ = url_.replace(/[?&]$/, "");

    let options_: AxiosRequestConfig = {
      method: "GET",
      url: url_,
      headers: {
        "Accept": "application/json"
      },
      cancelToken
    };

    return this.instance.request(options_).catch((_error: any) => {
      if (isAxiosError(_error) && _error.response) {
        return _error.response;
      } else {
        throw _error;
      }
    }).then((_response: AxiosResponse) => {
      return this.processUri(_response);
    });
}
````

Then you can (almost) safely use it in nuxt2 fetch helper like that:
````javascript
import axios from 'axios'

export default {
  async fetch({ store, params }) {
    const client = new FilesClient(process.env.BASE_URL, axios)
    const uri = await client.uri(params.id as string)
    store.dispatch('set-uri', uri)
  }
}
```` 

**How to update this flow to make it work with nuxt3/$fetch?**

**Reply by Kasheftin** [reply]
I found this nice memo on https://v3.nuxtjs.org/getting-started/data-fetching/

*You might be asking yourself: what is the difference between useFetch and useAsyncData? In brief, useFetch receives a URL and gets that data, whereas useAsyncData might have more complex logic. useFetch(url) is nearly equivalent to useAsyncData(url, () => $fetch(url)) - it's developer experience sugar for the most common use case.*

It helped a lot, at least I was able to use `useAsyncData` as a good old `fetch` from nuxt2:
````javascript
// store/users.ts:
import axios from 'axios'
import { defineStore } from 'pinia'

export const useUsersStore = defineStore({
  state: () => ({ users: [] }),
  actions: {
    async loadUsers() {
      const response = await axios.get('https://reqres.in/api/users?page=2')
      this.users = response.data.data
    }    
  }
})

// pages/page1.vue
<script setup lang="ts">
import { useUsersStore } from '@/store/users'

const usersStore = useUsersStore()

await useAsyncData(usersStore.loadUsers)
</script>
````

**Reply by steklopod** [reply]
The main reason why I am using axios is because  of`auth` module that requires it. By the way it is not, unfortunately,  yet working with nuxt 3 currently, so I have to use [@nuxtjs-alt/auth](https://github.com/Teranode/nuxt-module-alternatives) for now.

Also I use [openapi typescript-axios generator](https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/typescript-axios.md) for my nuxt 2 big project for generating frontend code from my kotlin-backend. There are only 2 generators for that:
- axios
- [fetch](https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/typescript-fetch.md)
And fetch generator does not have JWT support 😢 

<img width="514" alt="Снимок экрана 2022-09-23 в 17 37 06" src="https://user-images.githubusercontent.com/29500153/191998997-9b863deb-801c-4a19-84a3-d217b725bf1a.png">

**Reply by rikyotei** [reply]
same here, I'm using NSS to generateing openapi client code and really like to use axios. and an extra reason is fetch doesn't seem to support echoing progress when uploading files
